{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "67702266",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "forward primer length: 32\n",
      "reverse primer length: 32\n",
      "Total sequence number in database: 1061945\n",
      "Matched primer number: 43320\n",
      "primer match ratio: 0.04079307308758928\n",
      "----------------------------------------\n",
      "Matched target number: 24263\n",
      "target match ratio: 0.5600877192982456\n",
      "0.5600877192982456\n",
      "Matched target number: 24260\n",
      "target match ratio: 0.5600184672206833\n",
      "0.5600184672206833\n",
      "Matched target number: 24238\n",
      "target match ratio: 0.5595106186518929\n",
      "0.5595106186518929\n",
      "Matched target number: 24327\n",
      "target match ratio: 0.5615650969529086\n",
      "0.5615650969529086\n",
      "Matched target number: 23274\n",
      "target match ratio: 0.5372576177285319\n",
      "0.5372576177285319\n",
      "Matched target number: 23266\n",
      "target match ratio: 0.5370729455216989\n",
      "0.5370729455216989\n",
      "Matched target number: 24783\n",
      "target match ratio: 0.5720914127423823\n",
      "0.5720914127423823\n",
      "Matched target number: 21924\n",
      "target match ratio: 0.5060941828254848\n",
      "0.5060941828254848\n",
      "Matched target number: 3933\n",
      "target match ratio: 0.09078947368421053\n",
      "0.09078947368421053\n",
      "Matched target number: 5294\n",
      "target match ratio: 0.12220683287165282\n",
      "0.12220683287165282\n",
      "Matched target number: 4314\n",
      "target match ratio: 0.09958448753462604\n",
      "0.09958448753462604\n",
      "Matched target number: 4317\n",
      "target match ratio: 0.09965373961218836\n",
      "0.09965373961218836\n",
      "Matched target number: 4431\n",
      "target match ratio: 0.10228531855955679\n",
      "0.10228531855955679\n",
      "Matched target number: 4126\n",
      "target match ratio: 0.09524469067405356\n",
      "0.09524469067405356\n",
      "Matched target number: 4126\n",
      "target match ratio: 0.09524469067405356\n",
      "0.09524469067405356\n",
      "Matched target number: 4135\n",
      "target match ratio: 0.09545244690674054\n",
      "0.09545244690674054\n",
      "Matched target number: 4111\n",
      "target match ratio: 0.09489843028624192\n",
      "0.09489843028624192\n",
      "Matched target number: 4109\n",
      "target match ratio: 0.0948522622345337\n",
      "0.0948522622345337\n",
      "Matched target number: 4105\n",
      "target match ratio: 0.09475992613111726\n",
      "0.09475992613111726\n",
      "Matched target number: 3514\n",
      "target match ratio: 0.08111726685133887\n",
      "0.08111726685133887\n",
      "Matched target number: 3654\n",
      "target match ratio: 0.08434903047091413\n",
      "0.08434903047091413\n",
      "Matched target number: 3693\n",
      "target match ratio: 0.08524930747922438\n",
      "0.08524930747922438\n",
      "Matched target number: 2613\n",
      "target match ratio: 0.0603185595567867\n",
      "0.0603185595567867\n",
      "Matched target number: 2614\n",
      "target match ratio: 0.06034164358264081\n",
      "0.06034164358264081\n",
      "Matched target number: 2686\n",
      "target match ratio: 0.06200369344413666\n",
      "0.06200369344413666\n",
      "Matched target number: 2679\n",
      "target match ratio: 0.06184210526315789\n",
      "0.06184210526315789\n",
      "Matched target number: 2679\n",
      "target match ratio: 0.06184210526315789\n",
      "0.06184210526315789\n",
      "Matched target number: 2750\n",
      "target match ratio: 0.06348107109879964\n",
      "0.06348107109879964\n",
      "Matched target number: 14673\n",
      "target match ratio: 0.33871191135734074\n",
      "0.33871191135734074\n",
      "Matched target number: 14779\n",
      "target match ratio: 0.3411588180978763\n",
      "0.3411588180978763\n",
      "Matched target number: 20236\n",
      "target match ratio: 0.46712834718374885\n",
      "0.46712834718374885\n",
      "Matched target number: 19026\n",
      "target match ratio: 0.439196675900277\n",
      "0.439196675900277\n",
      "Matched target number: 13349\n",
      "target match ratio: 0.30814866112650047\n",
      "0.30814866112650047\n",
      "Matched target number: 13876\n",
      "target match ratio: 0.3203139427516159\n",
      "0.3203139427516159\n",
      "Matched target number: 13414\n",
      "target match ratio: 0.30964912280701756\n",
      "0.30964912280701756\n",
      "Matched target number: 13389\n",
      "target match ratio: 0.3090720221606648\n",
      "0.3090720221606648\n",
      "Matched target number: 13536\n",
      "target match ratio: 0.3124653739612188\n",
      "0.3124653739612188\n",
      "Matched target number: 13481\n",
      "target match ratio: 0.31119575253924286\n",
      "0.31119575253924286\n",
      "Matched target number: 13482\n",
      "target match ratio: 0.31121883656509697\n",
      "0.31121883656509697\n",
      "Matched target number: 18892\n",
      "target match ratio: 0.4361034164358264\n",
      "0.4361034164358264\n",
      "Matched target number: 18733\n",
      "target match ratio: 0.4324330563250231\n",
      "0.4324330563250231\n",
      "Matched target number: 18773\n",
      "target match ratio: 0.43335641735918745\n",
      "0.43335641735918745\n",
      "Matched target number: 21271\n",
      "target match ratio: 0.49102031394275164\n",
      "0.49102031394275164\n",
      "Matched target number: 20853\n",
      "target match ratio: 0.48137119113573407\n",
      "0.48137119113573407\n",
      "Matched target number: 20866\n",
      "target match ratio: 0.48167128347183746\n",
      "0.48167128347183746\n",
      "Matched target number: 21433\n",
      "target match ratio: 0.49475992613111724\n",
      "0.49475992613111724\n",
      "Matched target number: 17123\n",
      "target match ratio: 0.39526777469990765\n",
      "0.39526777469990765\n",
      "Matched target number: 16891\n",
      "target match ratio: 0.3899122807017544\n",
      "0.3899122807017544\n",
      "Matched target number: 17209\n",
      "target match ratio: 0.397253000923361\n",
      "0.397253000923361\n",
      "Matched target number: 15742\n",
      "target match ratio: 0.3633887349953832\n",
      "0.3633887349953832\n",
      "Matched target number: 15734\n",
      "target match ratio: 0.36320406278855033\n",
      "0.36320406278855033\n",
      "Matched target number: 16042\n",
      "target match ratio: 0.3703139427516159\n",
      "0.3703139427516159\n",
      "Matched target number: 21171\n",
      "target match ratio: 0.4887119113573407\n",
      "0.4887119113573407\n",
      "Matched target number: 22082\n",
      "target match ratio: 0.509741458910434\n",
      "0.509741458910434\n",
      "Matched target number: 23455\n",
      "target match ratio: 0.5414358264081256\n",
      "0.5414358264081256\n",
      "Matched target number: 23839\n",
      "target match ratio: 0.5503000923361034\n",
      "0.5503000923361034\n",
      "Matched target number: 23860\n",
      "target match ratio: 0.5507848568790397\n",
      "0.5507848568790397\n",
      "Matched target number: 24068\n",
      "target match ratio: 0.5555863342566943\n",
      "0.5555863342566943\n",
      "Matched target number: 23836\n",
      "target match ratio: 0.5502308402585411\n",
      "0.5502308402585411\n",
      "Matched target number: 23801\n",
      "target match ratio: 0.5494228993536473\n",
      "0.5494228993536473\n",
      "Matched target number: 23765\n",
      "target match ratio: 0.5485918744228994\n",
      "0.5485918744228994\n",
      "Matched target number: 23413\n",
      "target match ratio: 0.540466297322253\n",
      "0.540466297322253\n",
      "Matched target number: 23390\n",
      "target match ratio: 0.5399353647276085\n",
      "0.5399353647276085\n",
      "Matched target number: 24128\n",
      "target match ratio: 0.5569713758079409\n",
      "0.5569713758079409\n"
     ]
    }
   ],
   "source": [
    "# Screening HIV-1 sequences worldwide\n",
    "\n",
    "import RNA\n",
    "\n",
    "def DNA_reverse_complement_RNA(DNA):\n",
    "    complement = {'A': 'U', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "    return ''.join(complement.get(base, base) for base in reversed(DNA))\n",
    "\n",
    "# Only one base of permutation is tolerated and it should happen at the first two bases counting from the 3' end \n",
    "def single_base_permutation_3end_preserve_forward(DNA):\n",
    "    permutation_set = []\n",
    "    for i in range (1, len(DNA)-1):\n",
    "        permutation = DNA[:i-1] + 'A' + DNA[i:] \n",
    "        permutation_set.append(permutation)\n",
    "        permutation = DNA[:i-1] + 'C' + DNA[i:] \n",
    "        permutation_set.append(permutation)\n",
    "        permutation = DNA[:i-1] + 'G' + DNA[i:] \n",
    "        permutation_set.append(permutation)\n",
    "        permutation = DNA[:i-1] + 'T' + DNA[i:] \n",
    "        permutation_set.append(permutation)\n",
    "    return permutation_set\n",
    "\n",
    "def single_base_permutation_3end_preserve_reverse(DNA):\n",
    "    permutation_set = []\n",
    "    for i in range (4, len(DNA)+1):\n",
    "        permutation = DNA[:i-1] + 'A' + DNA[i:] \n",
    "        permutation_set.append(permutation)\n",
    "        permutation = DNA[:i-1] + 'C' + DNA[i:] \n",
    "        permutation_set.append(permutation)\n",
    "        permutation = DNA[:i-1] + 'G' + DNA[i:] \n",
    "        permutation_set.append(permutation)\n",
    "        permutation = DNA[:i-1] + 'T' + DNA[i:] \n",
    "        permutation_set.append(permutation)\n",
    "    return permutation_set\n",
    "\n",
    "# Only one base of insertion is tolerated, the insertion site should not be at the first two bases counting the 3' end or their closest neighbors  \n",
    "def single_base_insertion_3end_preserve_forward(DNA):\n",
    "    insertion_set = []\n",
    "    for i in range (1, len(DNA)-1):\n",
    "        insertion = DNA[:i-1] + 'A' + DNA[i-1:]\n",
    "        insertion_set.append(insertion)\n",
    "        insertion = DNA[:i-1] + 'C' + DNA[i-1:]\n",
    "        insertion_set.append(insertion)\n",
    "        insertion = DNA[:i-1] + 'G' + DNA[i-1:]\n",
    "        insertion_set.append(insertion)\n",
    "        insertion = DNA[:i-1] + 'T' + DNA[i-1:]\n",
    "        insertion_set.append(insertion)\n",
    "    return insertion_set\n",
    "        \n",
    "def single_base_insertion_3end_preserve_reverse(DNA):\n",
    "    insertion_set = []\n",
    "    for i in range (4, len(DNA)+1):\n",
    "        insertion = DNA[:i-1] + 'A' + DNA[i-1:]\n",
    "        insertion_set.append(insertion)\n",
    "        insertion = DNA[:i-1] + 'C' + DNA[i-1:]\n",
    "        insertion_set.append(insertion)\n",
    "        insertion = DNA[:i-1] + 'G' + DNA[i-1:]\n",
    "        insertion_set.append(insertion)\n",
    "        insertion = DNA[:i-1] + 'T' + DNA[i-1:]\n",
    "        insertion_set.append(insertion)\n",
    "    return insertion_set\n",
    "\n",
    "# Only one base of deletion is tolerated, the deleted base should not be the first two bases counting the 3' end or their closest neighbors  \n",
    "def single_base_deletion_3end_preserve_forward(DNA):\n",
    "    deletion_set = []\n",
    "    for i in range (1, len(DNA)-2):\n",
    "        deletion = DNA[:i-1] + DNA[i:]\n",
    "    return deletion_set\n",
    "    \n",
    "def single_base_deletion_3end_preserve_reverse(DNA):\n",
    "    deletion_set = []\n",
    "    for i in range (4, len(DNA)):\n",
    "        deletion = DNA[:i-1] + DNA[i:]\n",
    "    return deletion_set\n",
    "        \n",
    "ref_amplicon = 'AATGAGGAAGCTGCAGAATGGGATAGAGTGCATCCAGTGCATGCAGGGCCTATTGCACCAGGCCAGATGAGAGAACCAAGGGG'    # Amplicon sequence from the reference HIV-1 genome (NC001802.1)\n",
    "forward = 'AAGCAGCCATGCAAATGTTAAAAGAGACCATC'    # Forward RPA primer from the reference genome \n",
    "reverse = 'GAACCAAGGGGAAGTGACATAGCAGGAACTAC'    # Reverse RPA primer from the reference genome \n",
    "print('forward primer length: ' + str(len(forward)))\n",
    "print('reverse primer length: ' + str(len(reverse)))\n",
    "\n",
    "# Generating a primer set containing satisfying primers to be matched to sequencing records \n",
    "permutation_set_forward = single_base_permutation_3end_preserve_forward(forward)\n",
    "permutation_set_reverse = single_base_permutation_3end_preserve_reverse(reverse)\n",
    "insertion_set_forward = single_base_insertion_3end_preserve_forward(forward)\n",
    "insertion_set_reverse = single_base_insertion_3end_preserve_reverse(reverse)\n",
    "deletion_set_forward = single_base_deletion_3end_preserve_forward(forward)\n",
    "deletion_set_reverse = single_base_deletion_3end_preserve_reverse(reverse)\n",
    "\n",
    "# World wide HIV-1 sequencing records\n",
    "file = open('sequences-HIV-1-worldwide.fasta', 'r')  \n",
    "\n",
    "# North America HIV-1 sequencing records, remove the top and bottom semicolon to enable\n",
    "'''\n",
    "file = open('sequences-HIV-1-North America.fasta', 'r') \n",
    "'''\n",
    "\n",
    "# FASTA file processing\n",
    "lines = file.readlines()\n",
    "whole_seq = []\n",
    "for line in lines:\n",
    "    if line[0] == '>':\n",
    "        whole_seq.append('\\n')\n",
    "        whole_seq.append(line)\n",
    "        continue\n",
    "    else:\n",
    "        line = line.strip()\n",
    "        whole_seq.append(line)\n",
    "    \n",
    "whole_seq = ''.join(whole_seq)\n",
    "whole_seq = whole_seq[1:]\n",
    "\n",
    "whole_seq = whole_seq.splitlines()\n",
    "total_seq_num = int(len(whole_seq)/2)    # Return total number of sequencing records \n",
    "print('Total sequence number in database: ' + str(total_seq_num))\n",
    "\n",
    "# primer screen, the amplicon sequences with matched primers will be extracted for further processing\n",
    "amplicon_set = []\n",
    "primer_match_num = 0    # Initialize number of matched primer pairs \n",
    "\n",
    "for i in range (0, len(whole_seq)):\n",
    "    if i%2 == 1:\n",
    "        start = -1\n",
    "        end = -1\n",
    "        for j in range (0, len(whole_seq[i])-32+1):\n",
    "            if whole_seq[i][j:j+32] in permutation_set_forward:\n",
    "                start = j+32    # If yes, get the start base position\n",
    "                break\n",
    "        if start == -1:\n",
    "            for j in range (0, len(whole_seq[i])-33+1):\n",
    "                if whole_seq[i][j:j+33] in insertion_set_forward:\n",
    "                    start = j+33    # If yes, get the start base position\n",
    "                    break\n",
    "            if start == -1:\n",
    "                for j in range (0, len(whole_seq[i])-31+1):\n",
    "                    if whole_seq[i][j:j+31] in deletion_set_forward:\n",
    "                        start = j+31    # If yes, get the start base position\n",
    "                        break\n",
    "\n",
    "        for j in range (0, len(whole_seq[i])-32+1):\n",
    "            if whole_seq[i][j:j+32] in permutation_set_reverse:\n",
    "                end = j    # If yes, get the end base position\n",
    "                break\n",
    "        if start == -1:\n",
    "            for j in range (0, len(whole_seq[i])-33+1):\n",
    "                if whole_seq[i][j:j+33] in insertion_set_reverse:\n",
    "                    end = j    # If yes, get the end base position\n",
    "                    break\n",
    "            if start == -1:\n",
    "                for j in range (0, len(whole_seq[i])-31+1):\n",
    "                    if whole_seq[i][j:j+31] in deletion_set_reverse:\n",
    "                        end = j    # If yes, get the end base position\n",
    "                        break\n",
    "\n",
    "        if start != -1 and end != -1:\n",
    "            primer_match_num += 1    # If both start and end base position are confirmed, which means both the forward and reverse primer are satisfying, then store this sequencing record and move on to CRISPR target screening\n",
    "            insert = whole_seq[i][start:end+11]\n",
    "            amplicon_set.append(insert)\n",
    "\n",
    "\n",
    "print('Matched primer number: ' + str(primer_match_num))   # Return total number of sequencing records containing valid amplicons     \n",
    "ratio_1 = primer_match_num/total_seq_num\n",
    "print('primer match ratio: ' + str(ratio_1))    # Return the ratio of number of valid amplicons to the total number of sequencing records\n",
    "  \n",
    "print('----------------------------------------')    # Separator \n",
    "    \n",
    "# CRISPR target mismatch screening\n",
    "for s in range (0, len(ref_amplicon)-20+1):\n",
    "    ref_target = ref_amplicon[s:s+20]    # Target length fixed to 20 nt, can be customized \n",
    "\n",
    "    target_match_num = 0    # Initialize number of matched targets  \n",
    "\n",
    "    for i in range (0, len(amplicon_set)):\n",
    "        error_flag = 1    # If error_flag keeps 1 after screening, the target did not pass and should be discarded \n",
    "        for k in range (0, len(amplicon_set[i])-20+1):\n",
    "            if amplicon_set[i][k:k+20] == ref_target:\n",
    "                error_flag = -1    # error_flag changes to -1 if the target meets the screening criteria \n",
    "                break\n",
    "                        \n",
    "        if error_flag == -1:\n",
    "            target_match_num += 1\n",
    "\n",
    "    print('Matched target number: ' + str(target_match_num))   # Return number of matched targets\n",
    "    ratio_2 = target_match_num/primer_match_num    # Calculate the match ratio \n",
    "    print('target match ratio: ' + str(ratio_2))   # Return ratio of matched targets to the number of valid amplicons \n",
    "    print(ratio_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d5e8f06",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
